x/transfermiddleware/keeper/relay.go

```diff
@@ -0,0 +1,142 @@
+package keeper
+
+import (
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	transfertypes "github.com/cosmos/ibc-go/v7/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v7/modules/core/04-channel/types"
+	"github.com/notional-labs/banksy/v2/x/transfermiddleware/types"
+)
+
+func (k Keeper) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, data transfertypes.FungibleTokenPacketData) error {
+	if err := data.ValidateBasic(); err != nil {
+		return errorsmod.Wrapf(err, "error validating ICS-20 transfer packet data")
+	}
+	if !k.transferKeeper.GetReceiveEnabled(ctx) {
+		return transfertypes.ErrReceiveDisabled
+	}
+
+	// decode the receiver address
+	receiver, err := sdk.AccAddressFromBech32(data.Receiver)
+	if err != nil {
+		return errorsmod.Wrapf(err, "failed to decode receiver address: %s", data.Receiver)
+	}
+
+	// parse the transfer amount
+	transferAmount, ok := sdk.NewIntFromString(data.Amount)
+	if !ok {
+		return errorsmod.Wrapf(transfertypes.ErrInvalidAmount, "unable to parse transfer amount: %s", data.Amount)
+	}
+
+	paraTokenInfo := k.GetParachainIBCTokenInfo(ctx, data.Denom)
+
+	if transfertypes.ReceiverChainIsSource(packet.GetSourcePort(), packet.GetSourceChannel(), data.Denom) {
+		// Do nothing
+		return nil
+	}
+
+	// sender chain is the source, mint vouchers
```

Comment:
The voucher already minted and send to receiver in im.app.OnRecvPacket(ctx, packet, relayer). We can just send the minted token from receiver address to escrow address